src/entity/VanillaEntities.php

```diff
@@ -0,0 +1,154 @@
+	use RegistryTrait;
+
+	protected static function setup() : void{
+		self::_registryRegister("arrow", function(Location $location, ?Entity $shootingEntity, bool $critical, ?CompoundTag $nbt = null) : Arrow{
+			return new Arrow($location, $shootingEntity, $critical, $nbt);
+		});
+		self::_registryRegister("egg", function(Location $location, ?Entity $shootingEntity, ?CompoundTag $nbt = null) : Egg{
+			return new Egg($location, $shootingEntity, $nbt);
+		});
+		self::_registryRegister("ender_pearl", function(Location $location, ?Entity $shootingEntity, ?CompoundTag $nbt = null) : EnderPearl{
+			return new EnderPearl($location, $shootingEntity, $nbt);
+		});
+		self::_registryRegister("experience_bottle", function(Location $location, ?Entity $shootingEntity, ?CompoundTag $nbt = null) : ExperienceBottle{
+			return new ExperienceBottle($location, $shootingEntity, $nbt);
+		});
+		self::_registryRegister("experience_orb", function(Location $location, int $xpValue, ?CompoundTag $nbt = null) : ExperienceOrb{
+			return new ExperienceOrb($location, $xpValue, $nbt);
+		});
+		self::_registryRegister("falling_block", function(Location $location, Block $block, ?CompoundTag $nbt = null) : FallingBlock{
+			return new FallingBlock($location, $block, $nbt);
+		});
+		self::_registryRegister("item_entity", function(Location $location, Item $item, ?CompoundTag $nbt = null) : ItemEntity{
+			return new ItemEntity($location, $item, $nbt);
+		});
+		self::_registryRegister("painting", function(Location $location, Vector3 $blockIn, int $facing, PaintingMotive $motive, ?CompoundTag $nbt = null) : Painting{
+			return new Painting($location, $blockIn, $facing, $motive, $nbt);
+		});
+		self::_registryRegister("primed_tnt", function(Location $location, ?CompoundTag $nbt = null) : PrimedTNT{
+			return new PrimedTNT($location, $nbt);
+		});
+		self::_registryRegister("snowball", function(Location $location, ?Entity $shootingEntity, ?CompoundTag $nbt = null) : Snowball{
+			return new Snowball($location, $shootingEntity, $nbt);
+		});
+		self::_registryRegister("splash_potion", function(Location $location, ?Entity $shootingEntity, PotionType $potionType, ?CompoundTag $nbt = null) : SplashPotion{
+			return new SplashPotion($location, $shootingEntity, $potionType, $nbt);
+		});
+		self::_registryRegister("squid", function(Location $location, ?CompoundTag $nbt = null) : Squid{
+			return new Squid($location, $nbt);
+		});
+		self::_registryRegister("villager", function(Location $location, ?CompoundTag $nbt = null) : Villager{
+			return new Villager($location, $nbt);
+		});
+		self::_registryRegister("zombie", function(Location $location, ?CompoundTag $nbt = null) : Zombie{
+			return new Zombie($location, $nbt);
+		});
+		self::_registryRegister("human", function(Location $location, Skin $skin, ?CompoundTag $nbt = null) : Human{
+			return new Human($location, $skin, $nbt);
+		});
+	}
+
+	public static function register(string $name, \Closure $callback) : void{
```

Comment:
I don't like this. This is bringing back the exact same problems we had in the first place with inability to statically analyse code.

Reply:
I'm not sure how it could be declared in PHPStan that this only allows a compatible signature when compared to the entity registered before it

Reply:
It can't. That's exactly what I've been saying. This has the same problems that made the old systen unsuitable.